blockchain/chainio.go

```diff
@@ -1959,7 +1812,12 @@ func (b *BlockChain) initChainState(ctx context.Context) error {
 
 		return nil
 	})
-	return err
+	if err != nil {
+		return err
+	}
+
+	// Upgrade the database post block index load as needed.
+	return upgradeDBPostBlockIndexLoad(ctx, b)
```

Comment:
We discussed this a bit offline, but I'll comment here for future reference as well.

This will end up causing issues with future multi-version upgrades because it means that the upgrade process will not be able to complete uninterrupted and the need for it to run the normal code which relies on upgrades having happened.

For example, consider an upgrade from say version 7 to a future version 10.  It will complete the version 7 to 8 upgrade, but then not be able to get to version 9 because that can only happen once the normal upgrade process completes and the block index is fully loaded.  However, the code will otherwise depend on the database version having been upgraded to version 10 at that point, but it won't be.

We decided to go ahead and leave this in for now as the plan is to address it later by a more robust general reindexing solution.